theano/gpuarray/elemwise.py

```diff
@@ -108,15 +103,25 @@ def _generate_op_string(self, node):
         inps, outs = self._get_vnames(node)
         scal_v_ins = [get_scal(i.dtype)() for i in node.inputs]
 
-        fake_node = self.scalar_op.make_node(*scal_v_ins)
+        # As float16 isn't a c type and most GPU don't compute on it,
+        # We convert the computation to float32, and let libgpuarray
+        # load in float16 and cast to float32 and do the reverse for
+        # the output.
+        scalar_op = self.scalar_op
+        if isinstance(scalar_op, Composite):
+            s = scalar_op.clone_float32()
+            scalar_op = s
+        fake_node = scalar_op.make_node(*scal_v_ins)
         scal_v_out = fake_node.outputs
         assert len(scal_v_out) == len(node.outputs)
 
         kop = fake_node.op.c_code(fake_node, 'elem_scalar',
                                   inps, outs,
                                   dict(fail='return;'))
 
-        assert 'npy_float16' not in kop
+        if isinstance(scalar_op, Composite):
+            # Other op like cast should handle float16 themself.
+            assert 'npy_float16' not in kop
```

Comment:
So if you had just a cast op before this PR, how it was working? It didn't had special code to handle that before to my understanding. I think it still work (I'll add a test about that)